### PR TITLE
Raise CMake min to 3.14

### PR DESCRIPTION
**Pull request by jpr42**
- Explicitly test package manager support with CMake 3.14 and Ninja
- Remove gsl_install.cmake
- Simplify installation code

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -22,11 +22,28 @@ add_subdirectory(include)
...
+    set(namespace "Microsoft.GSL::")
+    set(cmake_files_install_dir ${CMAKE_INSTALL_DATADIR}/cmake/Microsoft.GSL)
+
+    install(TARGETS GSL EXPORT ${export_name} INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
+    install(EXPORT ${export_name} NAMESPACE ${namespace} DESTINATION ${cmake_files_install_dir})
+    export(TARGETS GSL NAMESPACE ${namespace} FILE ${export_name}.cmake)
+
+    set(gls_config_version "${CMAKE_CURRENT_BINARY_DIR}/Microsoft.GSLConfigVersion.cmake")
+
+    write_basic_package_version_file(${gls_config_version} COMPATIBILITY SameMajorVersion ARCH_INDEPENDENT)
```
Comment (jpr42) [inline CMakeLists.txt:46]: Reason for raising to 3.14
- To use ARCH_INDEPENDENT without condition (less for package managers to worry about)
- https://github.com/microsoft/GSL/issues/1084
- Simplifies writing github actions due to modern CLI (see cmake_find_package.yml)

File: cmake/gsl_install.cmake [hunk 2]
```diff
@@ -1,26 +0,0 @@
...
-# Install header files
-install(DIRECTORY "${PROJECT_SOURCE_DIR}/include/gsl" DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
-
-set(export_name "Microsoft.GSLConfig")
-set(namespace "Microsoft.GSL::")
-set(cmake_files_install_dir ${CMAKE_INSTALL_DATADIR}/cmake/Microsoft.GSL)
-
-# Add find_package() support
-target_include_directories(GSL INTERFACE $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
-install(TARGETS GSL EXPORT ${export_name})
```
Comment (jpr42) [inline cmake/gsl_install.cmake:13]: This had been updated to

```
install(TARGETS GSL EXPORT ${export_name} INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
```

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -1,13 +1,12 @@
...
 
 project(GSLTests LANGUAGES CXX)
-enable_testing()  # again, for support standalone testing
 
 set(GSL_CXX_STANDARD "14" CACHE STRING "Use c++ standard")
 
 set(CMAKE_CXX_STANDARD ${GSL_CXX_STANDARD})
 set(CMAKE_CXX_EXTENSIONS OFF)
-set(CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
Comment (jpr42) [inline tests/CMakeLists.txt:9]: minor mistake

File: tests/CMakeLists.txt [hunk 4]
```diff
@@ -57,11 +56,17 @@ if (NOT GTestMain_FOUND)
...
-    # import the globally installed Guidelines Support Library and test it
-    # instead of the current version from the include/ folder.
-    find_package(Microsoft.GSL REQUIRED)
+    find_package(Microsoft.GSL CONFIG REQUIRED)
+    enable_testing()
+
+    # Test version number support
+    if (NOT DEFINED Microsoft.GSL_VERSION)
+        message(FATAL_ERROR "Microsoft.GSL_VERSION not defined!")
+    endif()
```
Comment (jpr42) [inline tests/CMakeLists.txt:67]: Test the version number produced by `write_basic_package_version_file`

File: .github/workflows/cmake_find_package.yml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    name: Build ${{ matrix.os }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ ubuntu-20.04, ubuntu-22.04, macos-11, macos-12 ]
+    steps:
+      - uses: actions/checkout@v3
+      - uses: lukka/get-cmake@latest
+        with:
+          cmakeVersion: 3.14.0
```
Comment (jpr42) [inline .github/workflows/cmake_find_package.yml:20]: Check your cmake_minimum_required with `get-cmake`
https://discourse.cmake.org/t/check-your-cmake-minimum-required-with-get-cmake/7257

File: .github/workflows/cmake_find_package.yml [hunk 6]
```diff
@@ -0,0 +1,32 @@
+name: cmake_find_package
+
+on:
+    push:
+    pull_request:
+        branches:
+            - main
+
+jobs:
+  cmake-find-package:
```
Comment (dmitrykobets-msft) [inline .github/workflows/cmake_find_package.yml:10]: is this testing just `find_package` specifically, or the entire process of installing, building, and testing GSL?

File: tests/CMakeLists.txt [hunk 7]
```diff
@@ -57,11 +56,17 @@ if (NOT GTestMain_FOUND)
     )
 endif()
 
+# cmake_find_package.yml explicitly tests find_package support used by package managers
```
Comment (dmitrykobets-msft) [inline tests/CMakeLists.txt:59]: move comment to right above the `find_package` call?

File: tests/CMakeLists.txt [hunk 8]
```diff
@@ -57,11 +56,17 @@ if (NOT GTestMain_FOUND)
...
 if (CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
-    # CMake has been started independently in this directory with tests. Do
-    # import the globally installed Guidelines Support Library and test it
-    # instead of the current version from the include/ folder.
-    find_package(Microsoft.GSL REQUIRED)
+    find_package(Microsoft.GSL CONFIG REQUIRED)
+    enable_testing()
+
+    # Test version number support
+    if (NOT DEFINED Microsoft.GSL_VERSION)
```
Comment (dmitrykobets-msft) [inline tests/CMakeLists.txt:65]: this is coming from `find_package`, correct?

Comment (jpr42) [reply tests/CMakeLists.txt:65]: Correct.

The find_package command will create the variable `<PackageName>_VERSION`.
In this case <PackageName> is `Microsoft.GSL`
https://cmake.org/cmake/help/latest/command/find_package.html?highlight=find_package#config-mode-version-selection

File: .github/workflows/cmake_find_package.yml (hunk 6 above)
Comment (jpr42) [reply .github/workflows/cmake_find_package.yml:12]: 1.) It first tests installation.
2.) Tests the find_package support.
3.) It then builds/runs the tests to ensure the installation is 100% valid.

Part 3 is generally overkill but still helpful I think.

Also `actions` unlike `azure-pipelines` can be used in forks which can be helpful when working on a PR and not wanting to publish the PR until you are 100% ready.

File: .github/workflows/cmake_find_package.yml [hunk 9]
```diff
@@ -0,0 +1,32 @@
+name: cmake_find_package
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
```
Comment (jpr42) [inline .github/workflows/cmake_find_package.yml:7]: Updated to match `android.yml` and `ios.yml`

File: .github/workflows/cmake_find_package.yml (hunk 6 above)
Comment (dmitrykobets-msft) [reply .github/workflows/cmake_find_package.yml:10]: In that case, can this be renamed to something more general than `cmake_find_package.yaml`? This test seems to be performing an end-to-end test of installing GSL as a standalone project and running tests. Then

The comment `# cmake_find_package.yml explicitly tests find_package support used by package managers` inside `tests/CMakeLists.txt` emphasizes that this test is "explicitly" testing the `find_package` call but that just seems to be a by-product of the installation step.

Also out of curiosity, will replacing generator with something other than "Ninja" test anything different?

Comment (dmitrykobets-msft) [reply .github/workflows/cmake_find_package.yml:10]: (linking above comment to review)

Comment (jpr42) [reply .github/workflows/cmake_find_package.yml:10]: I've revised the script to just test `find_package` functionality.

> Also out of curiosity, will replacing generator with something other than "Ninja" test anything different?

For find_package testing I don't see a benefit.

**Comment by dmitrykobets-msft** [comment]
Great, thanks for the fix @jpr42

**Pull request merged**